Document behaviour of invalid YAML interleaved with valid

mthssdrbrg opened the pull request:
Follow up from #172.

Not entirely sure whether we actually want to return the valid objects, but maybe we do? We should probably mimic what `kubectl apply` does, and I think it fails if the file isn't valid YAML?

Commit:
Document behaviour of invalid YAML interleaved with valid

Not entirely sure whether we actually want to return the valid objects,
but maybe we do? We should probably mimic what `kubectl apply` does, and
I think it fails if the file isn't valid YAML?

JoelSpeed reviewed (dismissed):
Can you verify that this behaviour matches kubectl, otherwise happy to stick with that 👍

mthssdrbrg commented:
> Can you verify that this behaviour matches kubectl, otherwise happy to stick with that

Turns out it's actually not, it'll create all resources until it encounters an invalid one (`resources.yaml` contains three resources, second one is invalid):

```shell
$ kubectl apply -f resources.yaml
deployment.apps/nginx-valid created
error: error parsing resources.yaml: error converting YAML to JSON: yaml: invalid map key: map[interface {}]interface {}{"nginx--invalid":interface {}(nil)}
```

Commit:
Return partial results from `YAMLToUnstructured{Slice,}`

This is needed to be able to mimic the behaviour that `kubectl apply -f
<path>` has where it will parse and apply resources in a file until it
hits an error (and then continue with subsequent files if given a
directory).

Commit:
Mimic `kubectl apply -f <path>` behaviour in GitTrack controller

Previously if we encountered a file that we could partially parse we
would discard everything in said file, and eventually we would clean up
any "leftover" resources (which could be ones that we for some reason
now couldn't parse, perhaps because someone introduce a syntax error or
similar).

Now we instead try to partially apply what we can parse, which is what
`kubectl apply -f <path>` does as well. Hopefully this should be of
least surprise to users.

mthssdrbrg commented:
This has now been updated to mimic the behaviour of `kubectl apply -f <path>`, hopefully this provide users with a "least surprise" behaviour.

JoelSpeed commented on pkg/utils/decoder.go:
```diff
@@ -67,28 +67,29 @@ func YAMLToUnstructured(in []byte) (u unstructured.Unstructured, err error) {
 		if err != nil {
-			return []*unstructured.Unstructured{}, err
+			// unable to parse properly, bail
+			break
```
Any reason not to just return here? Since that's the effect of the break right?
```suggestion
			return us, err
```

If we do this then the error need not be scoped outside of the for loop which I think would probably be better

JoelSpeed reviewed (commented):
Added some suggestions as possible improvements, make of them what you will

Commit:
Prefer `SatisfyAll` over `And`

Commit:
Sprinkle some timeouts and `By` statements

Commit:
Simplify `YAMLToUnstructuredSlice` slightly

Avoid named returns as they're not as explicit as explicitly created
ones.

JoelSpeed reviewed (dismissed):
LGTM 👍

Commit:
Add test case for resources of invalid type

We're still not properly testing the function passed to `EachListItem`,
though I'm not sure if we actually can, I imagine it'll blow up
somewhere in-between transforming between YAML and JSON.

At least it covers one more error case.
